platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -1,73 +1,393 @@
-    let mut ctr = 0u32;
-    let mut linebuf = FixedVec::new((52 * 240) + 2).await;
-    for _ in 0..(52 * 240) + 2 {
-        let _ = linebuf.try_push(0);
+impl FullFrame {
+    pub fn new() -> Self {
+        Self {
+            frame: [[0u8; WIDTH_BYTES]; HEIGHT],
+        }
     }
+}
 
-    loop {
-        // It takes ~50ms to send a full frame, and 20fps is every
-        // 66.6ms. TODO: once we have "intervals", use that here.
-        k.sleep(Duration::from_millis(17)).await;
+impl FullFrame {
+    #[inline]
+    fn set_px(&mut self, x: usize, y: usize, color: Gray8) {
+        if x >= WIDTH || y > HEIGHT {
+            return;
+        }
+        let byte_x = x / 8;
+        let bit_x = x % 8;
 
-        // Send a pattern
-        //
-        // A note on this format:
-        //
-        // * Every FRAME gets a 1 byte command.
-        //     * It is zero, unless we are toggling VCOM.
-        //     * VCOM must be toggled once per second...ish.
-        // * Foreach LINE (240x) - 52 bytes total:
-        //     * 1 byte for line number
-        //     * (400bits / 8 = 50bytes) of data (one bit per pixel)
-        //     * 1 "dummy" byte
-        // * At the END, we need a total of two dummy bytes, so one from the last line, + 1 more
-        //
-        // This is where the 52x240 + 1 + 1 buffer size comes from.
+        if color.luma() > 128 {
+            self.frame[y][byte_x] |= 1 << (bit_x as u8);
+        } else {
+            self.frame[y][byte_x] &= !(1 << (bit_x as u8))
+        }
+    }
+}
+
+impl Dimensions for FullFrame {
+    fn bounding_box(&self) -> embedded_graphics::primitives::Rectangle {
+        Rectangle::new(
+            Point { x: 0, y: 0 },
+            Size {
+                width: 400,
+                height: 240,
+            },
+        )
+    }
+}
+
+impl DrawTarget for FullFrame {
+    type Color = Gray8;
+
+    type Error = ();
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        for px in pixels {
+            self.set_px(px.0.x as usize, px.0.y as usize, px.1)
+        }
+        Ok(())
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Helper tasks
+//
+// Friends to help us make things happen
+//////////////////////////////////////////////////////////////////////////////
+
+/// VCom - Once a second, update the vcom flag which is sent in every message.
+///
+/// TODO(AJM): The Beepberry uses a GPIO to toggle this rather than a bit in the
+/// SPI messages. This struct should be expanded to toggle the EXT pin at this
+/// rate instead, possibly as an optional field.
+struct VCom {
+    kernel: &'static Kernel,
+    ctxt: Arc<Mutex<Context>>,
+}
+
+impl VCom {
+    pub async fn run(self) {
+        loop {
+            self.kernel.sleep(Duration::from_secs(1)).await;
+            let mut c = self.ctxt.lock().await;
+            c.vcom = !c.vcom;
```

Comment:
in the future it seems like the vcom bit could maybe be an `AtomicBool`?

Reply:
Could be, though see the note above about supporting a gpio in the future